pkg/storegateway/bucket_stores.go

```diff
@@ -152,10 +153,31 @@ func (u *BucketStores) InitialSync(ctx context.Context) error {
 // SyncBlocks synchronizes the stores state with the Bucket store for every user.
 func (u *BucketStores) SyncBlocks(ctx context.Context) error {
 	return u.syncUsersBlocks(ctx, func(ctx context.Context, s *store.BucketStore) error {
-		return s.SyncBlocks(ctx)
+		return u.syncUsersBlocksWithRetries(ctx, s)
 	})
 }
 
+func (u *BucketStores) syncUsersBlocksWithRetries(ctx context.Context, s *store.BucketStore) error {
+	var lastErr error
+
+	retries := util.NewBackoff(ctx, util.BackoffConfig{
+		MinBackoff: 100 * time.Millisecond,
+		MaxBackoff: 10 * time.Second,
+		MaxRetries: 3,
+	})
+
+	for retries.Ongoing() {
+		lastErr = s.SyncBlocks(ctx)
+		if lastErr != nil {
```

Comment:
This will return error on first failure, and will not do any retry. Better version would be:

```
for retries.Ongoing() {
    err = s.SyncBlocks(ctx)
    if err == nil {
        return nil
    }
    // log error otherwise
    retries.Wait()
}
return retries.Err()
```